File: ios/Vidrip/ReaxnScreenRecorder.m
#import <React/RCTBridgeModule.h>
#import <ReplayKit/ReplayKit.h>
#import <AVFoundation/AVFoundation.h>

// Pure ObjC implementation so that @try/@catch around appendSampleBuffer is
// reliable. NSExceptions thrown inside Swift closures escape @try/@catch
// because Swift stack frames lack ObjC exception landing pads.

@interface ReaxnScreenRecorder : NSObject <RCTBridgeModule>
@end

@implementation ReaxnScreenRecorder {
  AVAssetWriter        *_writer;
  AVAssetWriterInput   *_videoIn;
  AVAssetWriterInput   *_audioIn;
  NSURL                *_outputURL;
  dispatch_queue_t      _writerQueue;
  BOOL                  _sessionStarted;
  volatile BOOL         _writingStarted;
  volatile BOOL         _cancelled;
  BOOL                  _isStarting;
}

RCT_EXPORT_MODULE()

+ (BOOL)requiresMainQueueSetup { return NO; }

// ---------------------------------------------------------------------------
#pragma mark - startCapture
// ---------------------------------------------------------------------------

RCT_EXPORT_METHOD(startCapture:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  RPScreenRecorder *recorder = [RPScreenRecorder sharedRecorder];
  if (recorder.isRecording || _isStarting) { resolve(nil); return; }

  _isStarting            = YES;
  recorder.microphoneEnabled = YES;

  NSString *name  = [NSString stringWithFormat:@"reaction_%@.mp4", [NSUUID UUID].UUIDString];
  NSURL    *url   = [[NSFileManager defaultManager].temporaryDirectory
                      URLByAppendingPathComponent:name];
  _outputURL      = url;
  _sessionStarted = NO;
  _writingStarted = NO;
  _cancelled      = NO;
  _writerQueue    = dispatch_queue_create("com.reaxn.assetwriter", DISPATCH_QUEUE_SERIAL);

  CGFloat scale = [UIScreen mainScreen].scale;
  CGSize  sz    = [UIScreen mainScreen].bounds.size;

  NSError *setupErr = nil;
  _writer = [AVAssetWriter assetWriterWithURL:url fileType:AVFileTypeMPEG4 error:&setupErr];
  if (setupErr) {
    _isStarting = NO;
    reject(@"SETUP_ERROR", setupErr.localizedDescription, setupErr);
    return;
  }

  _videoIn = [AVAssetWriterInput
    assetWriterInputWithMediaType:AVMediaTypeVideo
    outputSettings:@{
      AVVideoCodecKey:  AVVideoCodecTypeH264,
      AVVideoWidthKey:  @((NSInteger)(sz.width  * scale)),
      AVVideoHeightKey: @((NSInteger)(sz.height * scale)),
      AVVideoCompressionPropertiesKey: @{ AVVideoAverageBitRateKey: @4000000 },
    }];
  _videoIn.expectsMediaDataInRealTime = YES;

  _audioIn = [AVAssetWriterInput
    assetWriterInputWithMediaType:AVMediaTypeAudio
    outputSettings:@{
      AVFormatIDKey:         @(kAudioFormatMPEG4AAC),
      AVSampleRateKey:       @44100,
      AVNumberOfChannelsKey: @2,
      AVEncoderBitRateKey:   @128000,
    }];
  _audioIn.expectsMediaDataInRealTime = YES;

  [_writer addInput:_videoIn];
  [_writer addInput:_audioIn];

  __weak ReaxnScreenRecorder *weak = self;

  [recorder startCaptureWithHandler:^(CMSampleBufferRef sb,
                                       RPSampleBufferType type,
                                       NSError *captureErr) {
    ReaxnScreenRecorder *s = weak;
    if (!s || captureErr || !CMSampleBufferDataIsReady(sb)) return;
    if (!s->_writingStarted || s->_cancelled) return;

    // Drop early on the RP thread to avoid buffering frames in the serial queue.
    if (type == RPSampleBufferTypeVideo    && !s->_videoIn.isReadyForMoreMediaData) return;
    if (type == RPSampleBufferTypeAudioMic && !s->_audioIn.isReadyForMoreMediaData) return;
    if (type != RPSampleBufferTypeVideo && type != RPSampleBufferTypeAudioMic) return;

    // ReplayKit only guarantees the buffer is valid within the handler block,
    // so retain it before the async hop and release it after use.
    CFRetain(sb);

    dispatch_async(s->_writerQueue, ^{
      ReaxnScreenRecorder *ss = weak;
      if (!ss || ss->_cancelled) { CFRelease(sb); return; }
      if (ss->_writer.status != AVAssetWriterStatusWriting) { CFRelease(sb); return; }

      // Wrap startSession AND appendSampleBuffer — either can throw NSException.
      @try {
        if (!ss->_sessionStarted) {
          CMTime pts       = CMSampleBufferGetPresentationTimeStamp(sb);
          CMTime safeStart = CMTIME_IS_VALID(pts)
            ? CMTimeSubtract(pts, CMTimeMake(1, 10))
            : kCMTimeZero;
          [ss->_writer startSessionAtSourceTime:safeStart];
          ss->_sessionStarted = YES;
        }

        if (type == RPSampleBufferTypeVideo) {
          if (ss->_videoIn.isReadyForMoreMediaData) {
            [ss->_videoIn appendSampleBuffer:sb];
          }
        } else {
          if (ss->_audioIn.isReadyForMoreMediaData) {
            [ss->_audioIn appendSampleBuffer:sb];
          }
        }
      } @catch (NSException *) {
        ss->_cancelled = YES;
      }

      CFRelease(sb);
    });

  } completionHandler:^(NSError *startErr) {
    ReaxnScreenRecorder *s = weak;
    if (!s) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"DEALLOCATED", @"Recorder deallocated", nil);
      });
      return;
    }
    s->_isStarting = NO;
    if (startErr) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"START_ERROR", startErr.localizedDescription, startErr);
      });
      return;
    }
    if (![s->_writer startWriting]) {
      NSError *we = s->_writer.error;
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"WRITE_ERROR", we.localizedDescription ?: @"startWriting failed", we);
      });
      return;
    }
    s->_writingStarted = YES;
    dispatch_async(dispatch_get_main_queue(), ^{ resolve(nil); });
  }];
}

// ---------------------------------------------------------------------------
#pragma mark - stopCapture
// ---------------------------------------------------------------------------

RCT_EXPORT_METHOD(stopCapture:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  _cancelled = YES;

  if (![RPScreenRecorder sharedRecorder].isRecording) {
    resolve(@"");
    return;
  }

  __weak ReaxnScreenRecorder *weak = self;
  [[RPScreenRecorder sharedRecorder] stopCaptureWithHandler:^(NSError *err) {
    if (err) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"STOP_ERROR", err.localizedDescription, err);
      });
      return;
    }
    ReaxnScreenRecorder *s = weak;
    if (!s) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"NO_WRITER", @"Recorder deallocated", nil);
      });
      return;
    }
    // Drain writerQueue before finishing — ensures no append races markAsFinished.
    dispatch_async(s->_writerQueue, ^{
      [s->_videoIn markAsFinished];
      [s->_audioIn markAsFinished];
      [s->_writer finishWritingWithCompletionHandler:^{
        dispatch_async(dispatch_get_main_queue(), ^{
          NSString *path = s->_outputURL.path;
          if (path) { resolve(path); }
          else { reject(@"WRITE_ERROR", @"No output path", nil); }
        });
      }];
    });
  }];
}

// ---------------------------------------------------------------------------
#pragma mark - cancelCapture
// ---------------------------------------------------------------------------

RCT_EXPORT_METHOD(cancelCapture:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  _isStarting = NO;
  _cancelled  = YES;

  if (![RPScreenRecorder sharedRecorder].isRecording) {
    resolve(nil);
    return;
  }

  __weak ReaxnScreenRecorder *weak = self;
  [[RPScreenRecorder sharedRecorder] stopCaptureWithHandler:^(NSError *_) {
    ReaxnScreenRecorder *s = weak;
    dispatch_queue_t q = s ? s->_writerQueue : dispatch_get_main_queue();
    dispatch_async(q, ^{
      [s->_writer cancelWriting];
      if (s->_outputURL) {
        [[NSFileManager defaultManager] removeItemAtURL:s->_outputURL error:nil];
      }
      dispatch_async(dispatch_get_main_queue(), ^{ resolve(nil); });
    });
  }];
}

@end

File: ios/Reaxn/ReaxnScreenRecorder.swift
import Foundation
import ReplayKit
import AVFoundation

@objc(ReaxnScreenRecorder)
class ReaxnScreenRecorder: NSObject {

  private var writer: AVAssetWriter?
  private var videoIn: AVAssetWriterInput?
  private var audioIn: AVAssetWriterInput?
  private var outputURL: URL?
  private var sessionStarted = false

  // Serial queue — ensures startWriting() is never called twice from concurrent threads
  private let writeQueue = DispatchQueue(label: "com.reaxn.screenwriter", qos: .userInitiated)

  @objc static func requiresMainQueueSetup() -> Bool { return false }

  @objc func startCapture(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let recorder = RPScreenRecorder.shared()
    if recorder.isRecording {
      resolve(nil)
      return
    }

    recorder.isMicrophoneEnabled = true

    let ts = Int(Date().timeIntervalSince1970 * 1000)
    let url = FileManager.default.temporaryDirectory
      .appendingPathComponent("reaction_\(ts).mp4")
    self.outputURL = url
    self.sessionStarted = false

    let scale = UIScreen.main.scale
    let sz    = UIScreen.main.bounds.size
    let w     = Int(sz.width * scale)
    let h     = Int(sz.height * scale)

    do {
      let w_writer = try AVAssetWriter(outputURL: url, fileType: .mp4)

      let vi = AVAssetWriterInput(mediaType: .video, outputSettings: [
        AVVideoCodecKey:  AVVideoCodecType.h264,
        AVVideoWidthKey:  w,
        AVVideoHeightKey: h,
        AVVideoCompressionPropertiesKey: [AVVideoAverageBitRateKey: 4_000_000],
      ])
      vi.expectsMediaDataInRealTime = true

      let ai = AVAssetWriterInput(mediaType: .audio, outputSettings: [
        AVFormatIDKey:         kAudioFormatMPEG4AAC,
        AVSampleRateKey:       44100,
        AVNumberOfChannelsKey: 2,
        AVEncoderBitRateKey:   128_000,
      ])
      ai.expectsMediaDataInRealTime = true

      w_writer.add(vi)
      w_writer.add(ai)

      self.writer     = w_writer
      self.videoIn    = vi
      self.audioIn    = ai
    } catch let err {
      reject("SETUP_ERROR", err.localizedDescription, err)
      return
    }

    recorder.startCapture(handler: { [weak self] sb, type, err in
      // Route every buffer through the serial write queue — prevents concurrent
      // calls to startWriting() from racing each other and crashing AVAssetWriter
      self?.writeQueue.async { [weak self] in
        guard let strongSelf = self else { return }
        if let err = err {
          print("[ReaxnScreenRecorder] sample error: \(err)")
          return
        }
        guard CMSampleBufferDataIsReady(sb) else { return }

        if !strongSelf.sessionStarted {
          guard strongSelf.writer?.startWriting() == true else {
            print("[ReaxnScreenRecorder] startWriting failed: \(strongSelf.writer?.error?.localizedDescription ?? "unknown")")
            return
          }
          strongSelf.writer?.startSession(
            atSourceTime: CMSampleBufferGetPresentationTimeStamp(sb)
          )
          strongSelf.sessionStarted = true
        }

        switch type {
        case .video:
          if strongSelf.videoIn?.isReadyForMoreMediaData == true {
            strongSelf.videoIn?.append(sb)
          }
        case .audioMic:
          if strongSelf.audioIn?.isReadyForMoreMediaData == true {
            strongSelf.audioIn?.append(sb)
          }
        default:
          break
        }
      }
    }, completionHandler: { err in
      DispatchQueue.main.async {
        if let err = err {
          reject("START_ERROR", err.localizedDescription, err)
        } else {
          resolve(nil)
        }
      }
    })
  }

  @objc func stopCapture(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    RPScreenRecorder.shared().stopCapture { [weak self] err in
      if let err = err {
        DispatchQueue.main.async { reject("STOP_ERROR", err.localizedDescription, err) }
        return
      }
      guard let strongSelf = self else {
        DispatchQueue.main.async { reject("NO_WRITER", "Recorder was deallocated", nil) }
        return
      }
      // Finish writing on the serial queue to ensure all buffered appends are done
      strongSelf.writeQueue.async {
        strongSelf.videoIn?.markAsFinished()
        strongSelf.audioIn?.markAsFinished()
        strongSelf.writer?.finishWriting {
          DispatchQueue.main.async {
            if let path = strongSelf.outputURL?.path {
              resolve(path)
            } else {
              reject("WRITE_ERROR", "No output path", nil)
            }
          }
        }
      }
    }
  }

  @objc func cancelCapture(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    RPScreenRecorder.shared().stopCapture { [weak self] _ in
      self?.writeQueue.async {
        self?.writer?.cancelWriting()
        if let url = self?.outputURL {
          try? FileManager.default.removeItem(at: url)
        }
        DispatchQueue.main.async { resolve(nil) }
      }
    }
  }
}
